ArjunaCore/arjuna/tests/classes/com/hp/mwtests/ts/arjuna/atomicaction/TxStatsUnitTest.java

```diff
@@ -86,8 +84,42 @@ public void run() {
             t2.join();
         }
 
-        assertEquals(200, TxStats.getInstance().getNumberOfAbortedTransactions());
-        assertEquals(200, TxStats.getInstance().getNumberOfApplicationRollbacks());
-        assertEquals(401, TxStats.getInstance().getNumberOfTransactions());
+        Assertions.assertEquals(200, TxStats.getInstance().getNumberOfAbortedTransactions());
+        Assertions.assertEquals(200, TxStats.getInstance().getNumberOfApplicationRollbacks());
+        Assertions.assertEquals(400, TxStats.getInstance().getNumberOfTransactions());
+    }
+
+    @Test
+    @BMScript("TxStatsUnitTest/txStatsUnitTest")
+    public void inFlightTransactionsTest() {
+
+        // There shouldn't be any in-flight transaction now
+        Assertions.assertEquals(0, TxStats.getInstance().getNumberOfInflightTransactions());
+
+        createRendezvous();
+
+        Thread committingThread = new Thread(() -> {
```

Comment:
I don't agree with creating transactions in a separate thread to test the stat. I suspect that the reason you went for this option is that the existing test in this test class is leaving a transaction associated with the thread used to run your test. If you had created a separate class for testing the inflight transactions stat you would not have ran into this problem. The following test should be sufficient for testing the stat, which is easy to understand whereas the one in the PR is difficult to follow:

```
    @Test
    public void test2() {
        Assertions.assertNull(BasicAction.Current());

        Assertions.assertEquals(0, TxStats.getInstance().getNumberOfInflightTransactions());

        AtomicAction txn = new AtomicAction();

        txn.begin();

        Assertions.assertEquals(1, TxStats.getInstance().getNumberOfInflightTransactions());

        txn.commit();

        Assertions.assertEquals(0, TxStats.getInstance().getNumberOfInflightTransactions());
    }
```

I tried it and found that the initial assert will fail because the first test has left a transaction associated with the thread used to run the test. I think some routes through the code will leave abort transactions associated so that the caller can reason about the aborted transaction.

Reply:
Also, I thought you were going to test that the stat is reduced once the prepare phase begins and there is no test for that.

Reply:
> The existing tests are in TxStatsUnitTest but maybe add a new test class since testing in-flight vs preparing vs post prepared transactions might be tricky and make those tests harder to read, perhaps you could include a resource that delays the commit and/or the prepare while the test asks for the number of stats or some other way.

I must have misunderstood ^^^ request. TBH, it made sense that we wanted to check how `getNumberOfInflightTransactions()` works just before prepare. I will update this PR following your suggestions. Thanks

Reply:
> I tried it and found that the initial assert will fail because the first test has left a transaction associated with the thread used to run the test. I think some routes through the code will leave abort transactions associated so that the caller can reason about the aborted transaction.

This is a bug in the test. Found the solution. Should I fix it with a different JBTM? @mmusgrov

Reply:
> Also, I thought you were going to test that the stat is reduced once the prepare phase begins and there is no test for that.

Isn't this enough?
```
Assertions.assertEquals(1, TxStats.getInstance().getNumberOfInflightTransactions());
txn.commit();
Assertions.assertEquals(0, TxStats.getInstance().getNumberOfInflightTransactions());
```
> And a test that includes a test for rollback only transactions to verify that they aren't considered inflight.

Just to double check, you're talking about testing when the transaction is aborted, is that right?

Reply:
> > I tried it and found that the initial assert will fail because the first test has left a transaction associated with the thread used to run the test. I think some routes through the code will leave abort transactions associated so that the caller can reason about the aborted transaction.
> 
> This is a bug in the test. Found the solution. Should I fix it with a different JBTM? @mmusgrov

Just put your new test in its own test class and for now just include a comment in the existing TxStatsUnitTest test saying that it leaves the transaction associated. That way fixing the existing test will be low priority, and can happen at our own leisure, and will warn any future engineers to beware.

Reply:
I fixed the test already. I was just wondering if you prefer a new JBTM issue tracker for the bug or it is fine to include the fix in this PR

Reply:
> > Also, I thought you were going to test that the stat is reduced once the prepare phase begins and there is no test for that.
> 
> Isn't this enough?
> 
> ```
> Assertions.assertEquals(1, TxStats.getInstance().getNumberOfInflightTransactions());
> txn.commit();
> Assertions.assertEquals(0, TxStats.getInstance().getNumberOfInflightTransactions());
> ```
> 
> > And a test that includes a test for rollback only transactions to verify that they aren't considered inflight.
> 
> Just to double check, you're talking about testing when the transaction is aborted, is that right?

The check for inflight transactions is 

> > Also, I thought you were going to test that the stat is reduced once the prepare phase begins and there is no test for that.
> 
> Isn't this enough?
> 
> ```
> Assertions.assertEquals(1, TxStats.getInstance().getNumberOfInflightTransactions());
> txn.commit();
> Assertions.assertEquals(0, TxStats.getInstance().getNumberOfInflightTransactions());
> ```
> 
> > And a test that includes a test for rollback only transactions to verify that they aren't considered inflight.
> 
> Just to double check, you're talking about testing when the transaction is aborted, is that right?

The filter for inflight transactions is:

```
   entry.getValue().status() == ActionStatus.RUNNING &&
       entry.getValue().status() != ActionStatus.ABORT_ONLY)
```

so there should be a check that when the txn is in state ABORT_ONLY it will not be included in the stat and then
 when the txn starts preparing (ie when it makes the transition from RUNNING to ActionStatus.PREPARING) it similarly won't be included in the stat. Those are the only two extra conditions that I'd check for.

Reply:
But the stats test is effectively testing the state model so the maximal test would be to test the stats on every state transition and the minimal test would test for just RUNNING, ABORT_ONLY and PREPARING. I'm ok with us going for the minimal, but not less than that, test.

Reply:
Actually the test for inflight transactions is wrong,  `== ActionStatus.RUNNING && != ActionStatus.ABORT_ONLY` does not look correct - if it is RUNNING then && != ABORT_ONLY is vacuously true.

Reply:
I think` == ActionStatus.RUNNING` is enough to pick out an in flight transaction. Even though an ABORT_ONLY transaction has not started the 2PC protocol it is not considered to be active.

Reply:
> I fixed the test already. I was just wondering if you prefer a new JBTM issue tracker for the bug or it is fine to include the fix in this PR

Yes, do include it in the PR then.

Reply:
I hope I addressed all the above comments with my last (force) commit. I didn't go for another class as I used the old test method to add the checks you suggested. Please, let me know what you think about the new code. Thanks @mmusgrov

Reply:
Forgot to mention that I will squash the two commits together as soon as you approve this PR :-)